### PR TITLE
Refactor: Use item iterators for scroll checks

### PR DESCRIPTION
**Pull request by glebm**
Refs #2586

### CODE REVIEW

File: Source/control.cpp [hunk 1]
```diff
@@ -6,6 +6,7 @@
 #include "control.h"
 
 #include <array>
+#include <algorithm>
```
Comment (julealgon) [inline Source/control.cpp:9]: "al" comes before "ar"

File: Source/control.cpp [hunk 2]
```diff
@@ -684,6 +686,7 @@ void DrawSpellList(const Surface &out)
 	auto &myPlayer = Players[MyPlayerId];
 
 	for (auto &spellListItem : GetSpellListItems()) {
+		const spell_id spell = spellListItem.id;
```
Comment (julealgon) [inline Source/control.cpp:689]: Can we rename the variable to `spellId`?

File: Source/control.cpp [hunk 3]
```diff
@@ -723,22 +726,11 @@ void DrawSpellList(const Surface &out)
...
-				    && item._iSpell == spellListItem.id) {
-					v++;
-				}
-			}
-			strcpy(tempstr, fmt::format(ngettext("{:d} Scroll", "{:d} Scrolls", v), v).c_str());
+			const InventoryAndBeltPlayerItemsRange items { myPlayer };
+			const int count = std::count_if(items.begin(), items.end(), [spell](const Item &item) {
+				return item.IsScrollOf(spell);
+			});
+			strcpy(tempstr, fmt::format(ngettext("{:d} Scroll", "{:d} Scrolls", count), count).c_str());
```
Comment (julealgon) [inline Source/control.cpp:733]: Really nice cleanup.

File: Source/control.cpp [hunk 4]
```diff
@@ -1180,22 +1172,11 @@ void CheckPanelInfo()
...
-					if (!item.isEmpty()
-					    && (item._iMiscId == IMISC_SCROLL || item._iMiscId == IMISC_SCROLLT)
-					    && item._iSpell == v) {
-						s++;
-					}
-				}
-				strcpy(tempstr, fmt::format(ngettext("{:d} Scroll", "{:d} Scrolls", s), s).c_str());
+				const InventoryAndBeltPlayerItemsRange items { myPlayer };
+				const int count = std::count_if(items.begin(), items.end(), [v](const Item &item) {
+					return item.IsScrollOf(v);
```
Comment (julealgon) [inline Source/control.cpp:1177]: Can we rename `v` to `spellId` as well so that the changes in this PR are consistent and readable?

File: Source/control.cpp [hunk 5]
```diff
@@ -723,22 +726,11 @@ void DrawSpellList(const Surface &out)
...
-			for (auto &item : myPlayer.SpdList) {
-				if (!item.isEmpty()
-				    && (item._iMiscId == IMISC_SCROLL || item._iMiscId == IMISC_SCROLLT)
-				    && item._iSpell == spellListItem.id) {
-					v++;
-				}
-			}
-			strcpy(tempstr, fmt::format(ngettext("{:d} Scroll", "{:d} Scrolls", v), v).c_str());
+			const InventoryAndBeltPlayerItemsRange items { myPlayer };
+			const int count = std::count_if(items.begin(), items.end(), [spell](const Item &item) {
```
Comment (julealgon) [inline Source/control.cpp:730]: `count` is a bit generic for a name. Can we rename to something more semantic, like `scrollCount` maybe?

File: Source/inv.cpp [hunk 6]
```diff
@@ -1932,27 +1933,16 @@ bool UseScroll()
 	if (pcurs != CURSOR_HAND)
 		return false;
 
-	auto &myPlayer = Players[MyPlayerId];
+	Player &myPlayer = Players[MyPlayerId];
```
Comment (julealgon) [inline Source/inv.cpp:1936]: Why did you change from `auto` to `Player` here?

File: Source/inv_iterators.hpp [hunk 7]
```diff
@@ -19,7 +19,7 @@ class ItemsContainerRange {
 	class Iterator {
 	public:
 		using iterator_category = std::forward_iterator_tag;
-		using difference_type = void;
+		using difference_type = int;
```
Comment (julealgon) [inline Source/inv_iterators.hpp:22]: Can you elaborate why this is being changed? Was it wrong? Is it needed for the new iterator for some reason?

File: Source/inv_iterators.hpp [hunk 8]
```diff
@@ -261,6 +261,37 @@ class BeltPlayerItemsRange {
...
+		});
+	}
+
+	[[nodiscard]] ItemsContainerListRange::Iterator end() const // NOLINT(readability-identifier-naming)
+	{
+		auto result = ItemsContainerListRange::Iterator({
+		    InventoryPlayerItemsRange(*player_).end(),
+		    BeltPlayerItemsRange(*player_).end(),
+		});
+		return result;
```
Comment (julealgon) [inline Source/inv_iterators.hpp:288]: Can't we just return the iterator directly like we did in line 276? Would be good if for consistency alone.

File: Source/inv_iterators.hpp [hunk 9]
```diff
@@ -261,6 +261,37 @@ class BeltPlayerItemsRange {
...
+	}
+
+private:
+	Player *player_;
```
Comment (julealgon) [inline Source/inv_iterators.hpp:292]: Why does the field name end in an `_`? Shouldn't the underscore be _before_ the name to represent private data instead of _after_ it?

File: Source/items.h [hunk 10]
```diff
@@ -357,6 +357,16 @@ struct Item {
 		}
 	}
 
+	[[nodiscard]] bool IsScroll() const
```
Comment (julealgon) [inline Source/items.h:360]: Do we really need this `[[nodiscard]]`? What does it do?

File: Source/items.h [hunk 11]
```diff
@@ -357,6 +357,16 @@ struct Item {
 		}
 	}
 
+	[[nodiscard]] bool IsScroll() const
+	{
+		return _iMiscId == IMISC_SCROLL || _iMiscId == IMISC_SCROLLT;
```
Comment (julealgon) [inline Source/items.h:362]: Please use the `IsAnyOf` method.

File: Source/items.h [hunk 12]
```diff
@@ -357,6 +357,16 @@ struct Item {
 		}
 	}
 
+	[[nodiscard]] bool IsScroll() const
+	{
+		return _iMiscId == IMISC_SCROLL || _iMiscId == IMISC_SCROLLT;
+	}
+
+	[[nodiscard]] bool IsScrollOf(spell_id spell) const
```
Comment (julealgon) [inline Source/items.h:365]: Again would recommend renaming to `spellId`.

File: Source/player.cpp [hunk 13]
```diff
@@ -1973,17 +1974,9 @@ void CheckCheatStats(Player &player)
...
-		}
-	}
-
-	for (auto &item : SpdList) {
-		if (!item.isEmpty() && (item._iMiscId == IMISC_SCROLL || item._iMiscId == IMISC_SCROLLT)) {
-			if (item._iStatFlag)
-				_pScrlSpells |= GetSpellBitmask(item._iSpell);
+	for (Item &item : InventoryAndBeltPlayerItemsRange { *this }) {
+		if (item.IsScroll() && item._iStatFlag) {
+			_pScrlSpells |= GetSpellBitmask(item._iSpell);
```
Comment (julealgon) [inline Source/player.cpp:1979]: Can't we use something from algorithm to do this without an explicit foreach too (like we did for counting)?

File: Source/control.cpp (hunk 1 above)
Comment (glebm) [reply Source/control.cpp:8]: Done

File: Source/control.cpp (hunk 2 above)
Comment (glebm) [reply Source/control.cpp]: Done

File: Source/control.cpp (hunk 4 above)
Comment (glebm) [reply Source/control.cpp]: Done

File: Source/control.cpp (hunk 5 above)
Comment (glebm) [reply Source/control.cpp]: Done

File: Source/inv.cpp (hunk 6 above)
Comment (glebm) [reply Source/inv.cpp:1936]: On `auto`, from https://google.github.io/styleguide/cppguide.html#Type_deduction:

> Use type deduction only if it makes the code clearer to readers who aren't familiar with the project, or if it makes the code safer. Do not use it merely to avoid the inconvenience of writing an explicit type.

File: Source/inv_iterators.hpp (hunk 7 above)
Comment (glebm) [reply Source/inv_iterators.hpp:22]: `std::count_if` needs (and returns) `difference_type`

File: Source/inv_iterators.hpp (hunk 8 above)
Comment (glebm) [reply Source/inv_iterators.hpp]: Done

File: Source/items.h (hunk 12 above)
Comment (glebm) [reply Source/items.h]: Done

File: Source/inv_iterators.hpp (hunk 9 above)
Comment (glebm) [reply Source/inv_iterators.hpp:291]: After: https://github.com/diasurgical/devilutionX/blob/7771a08d033335e8857687dd0da258280908d345/Source/.clang-tidy#L70

File: Source/items.h (hunk 10 above)
Comment (glebm) [reply Source/items.h:360]: Errors if the return value is not used by the caller. Recommended for getter methods by clang-tidy.

File: Source/items.h (hunk 11 above)
Comment (glebm) [reply Source/items.h:362]: `IsAnyOf` seems to be a bit of an overkill in this case.

File: Source/player.cpp (hunk 13 above)
Comment (glebm) [reply Source/player.cpp:1979]: It is possible with `std::accumulate` but is less readable because we also filter and there isn't an `accumulate_if`

File: Source/inv.cpp (hunk 6 above)
Comment (julealgon) [reply Source/inv.cpp:1936]: Well that's Google's opinion right? Are we following _that_ standard for `auto`? I don't think so.

@AJenbo , can we enforce proper use of `auto` to avoid these discussions? Current codebase doesn't follow the rule @glebm just shared.

File: Source/items.h (hunk 10 above)
Comment (julealgon) [reply Source/items.h:360]: Interesting. I guess it would make sense to omit it only when there are out arguments and the user might potentially ignore the `bool` return and stuff like that.

Thanks for clarifying.

File: Source/items.h (hunk 11 above)
Comment (julealgon) [reply Source/items.h:362]: It's not a matter of being overkill, but of consistency.

Seems like @AJenbo himself marked as resolved... I don't get it :(

File: Source/inv.cpp (hunk 6 above)
Comment (StephenCWills) [reply Source/inv.cpp:1936]: It may not make the code clearer, but I also don't think it makes it any less clear. I would prefer `auto` here simply for consistency. All the cleanups recently have been using it in this same way.

Comment (julealgon) [reply Source/inv.cpp:1936]: > All the cleanups recently have been using it in this same way.

That's the most important aspect of my question: we've been doing tons of cleanup that includes `auto` uses like the one here, and now we are undoing some of it.

If we keep doing and undoing it that would be really confusing.

**Comment by glebm** [comment]
Whoa, rebasing is a bit of work :|

**Comment by glebm** [comment]
Can we merge this? Looks like all the discussions have been resolved.

**Comment by AJenbo** [comment]
Yes, it just needs a review :)

File: Source/items.h (hunk 11 above)
Comment (AJenbo) [reply Source/items.h:362]: Found it not important/optional, wanted a better overview

File: Source/player.cpp (hunk 13 above)
Comment (ephphatha) [reply Source/player.cpp:1979]: This could use a custom accumulator function, something like (untested):

```cpp
const InventoryAndBeltPlayerItemsRange items { *this };
_pScrlSpells = std::accumulate(items.begin(), items.end(), 0, [](uint64_t mask, const Item &item) {
    if (item.IsScroll() && item._iStatFlag) {
        mask |= GetSpellBitmask(item._iSpell);
    }
    return mask;
});
```

Don't think it's worth it for this case unless you really want to use `<algorithm>` concepts.

**Review by ephphatha — APPROVED** [review]
Looks good to me, left a comment on the accumulator discussion but the current implementation is good imo.
